### ISSUE TITLE
Provide access to Firefox's CDP implementation

### ISSUE THREAD

**Issue by Danp2** [Danp2/WebDriver#48]
From the latest geckodriver release notes --
```
Introduced the new boolean capability moz:debuggerAddress that can be used
to opt-in to the experimental Chrome DevTools Protocol (CDP) implementation.
A string capability with the same name will be returned by NewSession,
which contains the host:port combination of the HTTP server that can be
used to query for websockets of available targets.
```

Details on moz:debuggerAddress can be found [here](https://firefox-source-docs.mozilla.org/testing/geckodriver/Capabilities.html#moz-debuggeraddress).

**Comment by Danp2** [comment]
@TheDcoder Thoughts?

**Comment by TheDcoder** [comment]
Nice, but Mozilla is not exposing the interface itself via the WebDriver, they are using a separate instance with a different port. Their CDP interface is also different from Chrome, looks like they are providing even more low-level access as the API end-points they documented expose Web Sockets for communication.

TL;DR Our current implementation will not support this.

@Danp2 Thanks for digging this up though, maybe someone can implement this in the future :smile:

**Comment by Danp2** [comment]
Looks like WebSockets will be required to make this work with Firefox. See [here](https://groups.google.com/a/mozilla.org/g/dev-webdriver/c/DcYrCVCWMLU/m/U9V67ROGAQAJ) for their response.

Shame that there doesn't appear to be a good implementation in AutoIt.

**Comment by Danp2** [comment]
@TheDcoder What about using Javascript to initiate / control the websocket instead of using WinHTTP?

**Comment by TheDcoder** [comment]
I don't think we can do that, I assume you are referring to JS executed via `ExecuteScript`, it simply doesn't have access to connect to local or 3rd-party domains since it is injected into the webpage's context.

**Comment by Danp2** [comment]
I was thinking that `_WD_ExecuteScript` could be used to execute JS to create the WS and store it in a global variable. My thinking is that if I can do it from the webbrowser's developer console, then I should be able to do it via webdriver.

**Comment by TheDcoder** [comment]
> if I can do it from the webbrowser's developer console

Did you test this? I doubt this would work.

**Comment by Danp2** [comment]
Not yet, but I will soon.

**Comment by Danp2** [comment]
@TheDcoder Looks like you were correct about the websocket creation now working due to security restrictions. However, it looks like I may have come up with a generic function to return various CDP-related settings. Please take a look at #82 and provide feedback.

P.S. I have the WS piece working in Autoit as well ;-)

**Comment by TheDcoder** [comment]
Nice, but I am not really familiar with these interfaces and I don't have a ready to test rig set up for browser automation.

> P.S. I have the WS piece working in Autoit as well ;-)

Now that's cool, mind sharing your secret WS UDF? :yum:

**Comment by Danp2** [comment]
See https://github.com/Danp2/autoit-websocket. It's a slightly modified version of code posted by Firefox [here](https://www.autoitscript.com/forum/topic/84133-winhttp-functions/page/30/?tab=comments#comment-1172353).
